### PR TITLE
BLACK 529 -- Filter functions

### PR DESCRIPTION
**Pull request by saltysealion**
+ Add filter functions

### CODE REVIEW

File: tests/test_filter_functions.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        system=system
+    )
+
+    noise = AdditiveNoiseFactory.create(
+        name='Dephasing',
+        operator=SIGMA_Z,
+        system=system
+    )
+
+    FilterFunctionFactory.create(
```
Comment (stabacco) [inline tests/test_filter_functions.py]: should you be passing the noise here ?

File: qctrl/qctrl.py [hunk 2]
```diff
@@ -13,29 +13,29 @@
...
 
-
-# @pylint: disable=too-few-public-methods
+# @pylint: disable=too-few-public-methods, unnecessary-lambda
 class Qctrl:
     """Managing Q-CTRL controls.
     """
 
-    _factory_registry = FACTORY_REGISTRY
+    _service_registry = SERVICE_REGISTRY
```
Comment (stabacco) [inline qctrl/qctrl.py]: these should use the `ClassRegistry` as well now

File: qctrl/qctrl.py [hunk 3]
```diff
@@ -49,22 +49,33 @@ def __init__(self, auth_token: str,
...
-            setattr(self, factory, instance)
+            setattr(self.services, service, instance)
 
         # expose api models
+        self.models = lambda: None
         methods = [attribute for attribute in inspect.getmembers(self.api) \
             if isinstance(attribute[1], model.Model)]
         for (model_name, instance) in methods:
-            setattr(self, model_name, instance)
+            model_class_name = re.sub(r'\w+', lambda m: m.group(0).capitalize(), model_name)
```
Comment (stabacco) [inline qctrl/qctrl.py]: can you add a bit of comment about what’s going on here?

File: qctrl/schema.py [hunk 4]
```diff
@@ -19,8 +19,8 @@
 import numpy as np
```
Comment (stabacco) [inline qctrl/schema.py]: is the schema still being used? I thought the factories would deprecate this.

File: qctrl/services/__init__.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+models as inputs and are responsible for orchestrating the execution of the
+task the user has invoked. Services accept and return models.
+"""
+from .optimizations import OptimalOptimizedSystem, RobustOptimizedSystem
+from .metrics import T1DecayAssessment
+
+
+# The REGISTRY of all available optimizations, it will be used by the
+# Qctrl session object.
+REGISTRY = {
```
[the remark this answers is not in the document]
Comment (saltysealion) [reply qctrl/services/__init__.py]: Lets do this in a future revision

File: tests/test_filter_functions.py (hunk 1 above)
Comment (saltysealion) [reply tests/test_filter_functions.py]: Yes!

File: qctrl/qctrl.py (hunk 3 above)
Comment (saltysealion) [reply qctrl/qctrl.py]: Sure

File: qctrl/schema.py (hunk 4 above)
Comment (saltysealion) [reply qctrl/schema.py]: I wish I could delete this, but it's still being used for dump

**Pull request merged**